crates/ra_hir/src/impl_block.rs

```diff
@@ -0,0 +1,169 @@
+        file_items: &SourceFileItems,
+        module: &Module,
+        node: ast::ImplBlock,
+    ) -> Self {
+        let impl_for = TypeRef::from_ast_opt(node.target_type());
+        let items = if let Some(item_list) = node.item_list() {
+            item_list
+                .impl_items()
+                .map(|item_node| {
+                    let kind = match item_node {
+                        ast::ImplItem::FnDef(..) => DefKind::Function,
+                        ast::ImplItem::ConstDef(..) => DefKind::Item,
+                        ast::ImplItem::TypeDef(..) => DefKind::Item,
+                    };
+                    let item_id = file_items.id_of_unchecked(item_node.syntax());
+                    let def_loc = DefLoc {
+                        kind,
+                        source_root_id: module.source_root_id,
+                        module_id: module.module_id,
+                        source_item_id: SourceItemId {
+                            file_id: file_items.file_id,
+                            item_id: Some(item_id),
+                        },
+                    };
+                    let def_id = def_loc.id(db);
+                    match item_node {
+                        ast::ImplItem::FnDef(..) => ImplItem::Method(Function::new(def_id)),
+                        ast::ImplItem::ConstDef(..) => ImplItem::Const(def_id),
+                        ast::ImplItem::TypeDef(..) => ImplItem::Type(def_id),
+                    }
+                })
+                .collect()
+        } else {
+            Vec::new()
+        };
+        ImplData { impl_for, items }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum ImplItem {
+    Method(Function),
+    // these don't have their own types yet
+    Const(DefId),
+    Type(DefId),
+    // Existential
+}
+
+impl ImplItem {
+    pub fn def_id(&self) -> DefId {
+        match self {
+            ImplItem::Method(f) => f.def_id(),
+            ImplItem::Const(def_id) => *def_id,
+            ImplItem::Type(def_id) => *def_id,
+        }
+    }
+}
+
+pub type ImplId = Id<ImplData>;
+
+/// We have to collect all impl blocks in a crate, to later be able to find
+/// impls for specific types.
+#[derive(Debug, PartialEq, Eq)]
+pub struct CrateImplBlocks {
+    impls: Arena<ImplData>,
+    impls_by_def: FxHashMap<DefId, ImplId>,
+}
+
+impl CrateImplBlocks {
+    fn new() -> Self {
+        CrateImplBlocks {
+            impls: Arena::new(),
+            impls_by_def: FxHashMap::default(),
+        }
+    }
+
+    fn collect(&mut self, db: &impl HirDatabase, module: Module) -> Cancelable<()> {
+        let module_source_node = module.source().resolve(db);
+        let node = match &module_source_node {
+            ModuleSourceNode::SourceFile(node) => node.borrowed().syntax(),
+            ModuleSourceNode::Module(node) => node.borrowed().syntax(),
+        };
+
+        let source_file_items = db.file_items(module.source().file_id());
```

Comment:
`file_items` its unfortunate query to have as a dependency. It contains `SyntaxTrees` directly, so any change to the file invalidates `file_items`. 

With the current setup, `impls_in_crate` query depends **directly** on all `file_items` of all files. That means that any change will cause `impls_in_crate` query to be recomputed on every change. And because it can't just update changes to a single module, it will effectively traverse all the files in the crate.

This I think is a "salsa smell": although the results of `impls_in_crate` are expected to be stable across the majority of edits, we recompute them after each change!

To deal with this issue, we need to add an intermediate query `impls_in_file` which calls `file_items`, extracts impls and strips the infor about `SyntaxNodes`. Because `impls_in_file` will store only `SourceFileItemId`s, which are stable across reparses, we get a much better reuse. After a file is edited, `impls_in_file` for **this** file will be recomputed and compared with the old value. However in the common case the result would be the same, and so `impls_in_crate` **won't** be recomputed. 

This means that we pay for incrementality with a space overhead: each impl will be stored in file-local query, and in crate-global query (not unlike segment trees/monoid cached trees). Ideally, we'll store intermediate results for the *current* crate and gc intermediaries for the readonly crates.io crate. 

Note that a very similar thing happens when computing `ModuleItemMap`: there's an intermediate per-module `input_module_items` which depends on `file_items` and "projects" `SourceFileItemId` out of `(SourceFileItemId, SyntaxNode)` pairs. I think we should employ a similar pattern here (and perhaps even reuse `InputModuleItems` for both impls and name res?).

We also should have a test for "local typing does not invalidate world", like this one:

https://github.com/rust-analyzer/rust-analyzer/blob/c0d1b17a4e976e07966a1d7f545b6e01d278fd08/crates/ra_hir/src/module/nameres/tests.rs#L143

Reply:
Another instance of this pattern is `submodules` and `module_tree` queries.

Reply:
> This means that we pay for incrementality with a space overhead: each impl will be stored in file-local query, and in crate-global query (not unlike segment trees/monoid cached trees). 
I think we could just keep them in the file-local query, and have the crate-global query just refer to the file and impl ID?

Reply:
Yep, that also will work, although perhaps `ModuleId` should be included as well (a single `FileId` might correspond to several modules in different crate or even in a single crate).